src/pymor/algorithms/timestepping.py

```diff
@@ -93,17 +153,80 @@ class ImplicitEulerTimeStepper(TimeStepper):
     def __init__(self, nt, solver_options='operator'):
         self.__auto_init(locals())
 
-    def solve(self, initial_time, end_time, initial_data, operator, rhs=None, mass=None, mu=None, num_values=None):
-        return implicit_euler(operator, rhs, mass, initial_data, initial_time, end_time, self.nt, mu, num_values,
-                              solver_options=self.solver_options)
+    def estimate_time_step_number(self, initial_time, end_time):
+        return self.nt + 1
+
+    def iterate(self, initial_time, end_time, initial_data, operator, rhs=None, mass=None, mu=None, num_values=None):
+        A, F, M, U0, t0, t1, nt = operator, rhs, mass, initial_data, initial_time, end_time, self.nt
+        assert isinstance(A, Operator)
+        assert isinstance(F, (type(None), Operator, VectorArray))
+        assert isinstance(M, (type(None), Operator))
+        assert A.source == A.range
+        num_values = num_values or nt + 1
+        dt = (t1 - t0) / nt
+        DT = (t1 - t0) / (num_values - 1)
+
+        if F is None:
+            F_time_dep = False
+        elif isinstance(F, Operator):
+            assert F.source.dim == 1
+            assert F.range == A.range
+            F_time_dep = _depends_on_time(F, mu)
+            if not F_time_dep:
+                dt_F = F.as_vector(mu) * dt
+        else:
+            assert len(F) == 1
+            assert F in A.range
+            F_time_dep = False
+            dt_F = F * dt
+
+        if M is None:
+            from pymor.operators.constructions import IdentityOperator
+            M = IdentityOperator(A.source)
+
+        assert A.source == M.source == M.range
+        assert not M.parametric
+        assert U0 in A.source
+        assert len(U0) == 1
+
+        time_steps = 1
+        yield U0, t0
+
+        options = (A.solver_options if self.solver_options == 'operator' else
+                   M.solver_options if self.solver_options == 'mass' else
+                   self.solver_options)
+        M_dt_A = (M + A * dt).with_(solver_options=options)
+        if not _depends_on_time(M_dt_A, mu):
+            M_dt_A = M_dt_A.assemble(mu)
+
+        t = t0
+        U = U0.copy()
+        if mu is None:
+            mu = Mu()
+
+        for n in range(nt):
+            t += dt
+            mu = mu.with_(t=t)
+            rhs = M.apply(U)
+            if F_time_dep:
+                dt_F = F.as_vector(mu) * dt
+            if F:
+                rhs += dt_F
+            U = M_dt_A.apply_inverse(rhs, mu=mu, initial_guess=U)
+            while t - t0 + (min(dt, DT) * 0.5) >= time_steps * DT:
+                time_steps += 1
```

Comment:
Regarding our discussion above: For `DT < dt` this would be the number of returned values, not the number of time steps ..